File: app/src/main/java/com/wavecat/inline/preferences/LinearGroup.java
package com.wavecat.inline.preferences;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.SharedPreferences;
import android.view.View;
import android.widget.LinearLayout;

import org.luaj.vm2.LuaTable;
import org.luaj.vm2.LuaValue;
import org.luaj.vm2.Varargs;

@SuppressLint("ViewConstructor")
public class LinearGroup extends LinearLayout implements Preference {

    private final LuaTable views;

    public LinearGroup(Context context, LuaTable views) {
        super(context);

        this.views = views;
    }

    @Override
    public View getView(SharedPreferences preferences) {
        removeAllViews();

        LuaValue k = LuaValue.NIL;
        while (true) {
            Varargs n = views.next(k);

            if ((k = n.arg1()).isnil())
                break;

            addView(Utils.castPreference(getContext(), n.arg(2)).getView(preferences));
        }

        return this;
    }
}

File: app/src/main/java/com/wavecat/inline/preferences/views/LinearGroup.kt
package com.wavecat.inline.preferences.views

import android.annotation.SuppressLint
import android.content.Context
import android.content.SharedPreferences
import android.view.View
import android.widget.LinearLayout
import com.wavecat.inline.extensions.forEach
import com.wavecat.inline.preferences.Preference
import com.wavecat.inline.preferences.castPreference
import org.luaj.vm2.LuaTable

/**
 * A [LinearLayout] that displays a group of preferences.
 * The preferences are provided as a [LuaTable] and are added to the layout dynamically.
 *
 * @param context The context in which the view is created.
 * @param views A [LuaTable] containing the preferences to be displayed.
 */
@SuppressLint("ViewConstructor")
open class LinearGroup(context: Context, private val views: LuaTable) : LinearLayout(context), Preference {
    override fun getView(preferences: SharedPreferences?, requestFocus: () -> Unit): View {
        removeAllViews()

        views.forEach { _, value ->
            addView(
                castPreference(
                    context = context,
                    value = value,
                    topOrientation = orientation
                )
                    .getView(preferences, requestFocus)
            )
        }

        return this
    }
}
